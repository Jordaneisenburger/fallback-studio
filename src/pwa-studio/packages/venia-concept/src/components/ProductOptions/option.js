import React, { Component } from 'react';
import { arrayOf, func, object, shape, string } from 'prop-types';

import classify from '../../classify';
import getOptionType from './getOptionType';
import SwatchList from './swatchList';
import TileList from './tileList';
import defaultClasses from './option.css';

const getItemKey = ({ value_index }) => value_index;

class Option extends Component {
    static propTypes = {
        attribute_id: string,
        attribute_code: string.isRequired,
        classes: shape({
            root: string,
            title: string
        }),
        label: string.isRequired,
        onSelectionChange: func,
        values: arrayOf(object).isRequired
    };

    handleSelectionChange = selection => {
        const { attribute_id, onSelectionChange } = this.props;

        if (onSelectionChange) {
            onSelectionChange(attribute_id, selection);
        }
    };

    get listComponent() {
        const { attribute_code, values } = this.props;

        // TODO: get an explicit field from the API
        // that identifies an attribute as a swatch
        const optionType = getOptionType({ attribute_code, values });

        return optionType === 'swatch' ? SwatchList : TileList;
    }

    render() {
        const { handleSelectionChange, listComponent: ValueList, props } = this;
        const { classes, label, values } = props;

        return (
            <div className={classes.root}>
                <h3 className={classes.title}>
                    <span>{label}</span>
                </h3>
                <ValueList
                    getItemKey={getItemKey}
                    items={values}
                    onSelectionChange={handleSelectionChange}
                />
            </div>
        );
    }
}

export default classify(defaultClasses)(Option);
