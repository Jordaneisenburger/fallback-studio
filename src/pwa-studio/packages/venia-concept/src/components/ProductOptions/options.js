import React, { Component } from 'react';
import { func, object } from 'prop-types';

import isProductConfigurable from 'src/util/isProductConfigurable';

import Option from './option';

class Options extends Component {
    static propTypes = {
        onSelectionChange: func,
        product: object
    };

    handleSelectionChange = (optionId, selection) => {
        const { onSelectionChange } = this.props;

        if (onSelectionChange) {
            onSelectionChange(optionId, selection);
        }
    };

    render() {
        const { handleSelectionChange, props } = this;
        const { product } = props;

        if (!isProductConfigurable(product)) {
            // Non-configurable products don't have options.
            return null;
        }

        const { configurable_options } = product;
        return configurable_options.map(option => (
            <Option
                {...option}
                key={option.attribute_id}
                onSelectionChange={handleSelectionChange}
            />
        ));
    }
}

export default Options;
