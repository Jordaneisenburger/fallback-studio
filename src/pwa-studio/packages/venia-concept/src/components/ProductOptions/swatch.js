import React, { Component } from 'react';
import { bool, number, object, oneOfType, shape, string } from 'prop-types';

import classify from 'src/classify';
import Icon from 'src/components/Icon';
import Tooltip from './toolTip';
import CheckIcon from 'react-feather/dist/icons/check';

import defaultClasses from './swatch.css';

import { memoizedGetRandomColor } from 'src/util/getRandomColor';

const getClassName = (name, isSelected, hasFocus) =>
    `${name}${isSelected ? '_selected' : ''}${hasFocus ? '_focused' : ''}`;

class Swatch extends Component {
    static propTypes = {
        classes: shape({
            root: string
        }),
        hasFocus: bool,
        isSelected: bool,
        item: shape({
            label: string.isRequired,
            value_index: oneOfType([number, string]).isRequired
        }).isRequired,
        itemIndex: number,
        style: object
    };

    static defaultProps = {
        hasFocus: false,
        isSelected: false
    };

    get icon() {
        const { isSelected } = this.props;

        return isSelected ? <Icon src={CheckIcon} /> : null;
    }

    render() {
        const { icon, props } = this;
        const {
            classes,
            hasFocus,
            isSelected,
            item,
            // eslint-disable-next-line
            itemIndex,
            style,
            ...restProps
        } = props;

        const className = classes[getClassName('root', isSelected, hasFocus)];
        const { label, value_index } = item;

        // TODO: use the colors from graphQL when they become available.
        const randomColor = memoizedGetRandomColor(value_index);

        // We really want to avoid specifying presentation within JS.
        // Swatches are unusual in that their color is data, not presentation,
        // but applying color *is* presentational.
        // So we merely provide the color data here, and let the CSS decide
        // how to use that color (e.g., background, border).
        const finalStyle = Object.assign({}, style, {
            '--venia-swatch-bg': randomColor
        });

        return (
            <Tooltip text={label}>
                <button
                    {...restProps}
                    className={className}
                    style={finalStyle}
                    title={label}
                >
                    {icon}
                </button>
            </Tooltip>
        );
    }
}

export default classify(defaultClasses)(Swatch);
