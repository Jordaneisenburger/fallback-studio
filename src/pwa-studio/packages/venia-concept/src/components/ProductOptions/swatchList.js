import React, { Component } from 'react';
import { arrayOf, object, shape, string } from 'prop-types';
import { List } from '@magento/peregrine';

import classify from '../../classify';
import Swatch from './swatch';
import defaultClasses from './swatchList.css';

class SwatchList extends Component {
    static propTypes = {
        classes: shape({
            root: string
        }),
        items: arrayOf(object)
    };

    render() {
        return <List renderItem={Swatch} {...this.props} />;
    }
}

export default classify(defaultClasses)(SwatchList);
