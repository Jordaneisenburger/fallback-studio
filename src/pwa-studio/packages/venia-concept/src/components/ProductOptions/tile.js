import React, { Component } from 'react';
import { bool, number, shape, string } from 'prop-types';

import classify from '../../classify';
import Tooltip from './toolTip';
import defaultClasses from './tile.css';

const getClassName = (name, isSelected, hasFocus) =>
    `${name}${isSelected ? '_selected' : ''}${hasFocus ? '_focused' : ''}`;

class Tile extends Component {
    static propTypes = {
        classes: shape({
            root: string
        }),
        hasFocus: bool,
        isSelected: bool,
        item: shape({
            label: string.isRequired
        }).isRequired,
        itemIndex: number
    };

    static defaultProps = {
        hasFocus: false,
        isSelected: false
    };

    render() {
        const {
            classes,
            hasFocus,
            isSelected,
            item,
            // eslint-disable-next-line
            itemIndex,
            ...restProps
        } = this.props;
        const className = classes[getClassName('root', isSelected, hasFocus)];
        const { label } = item;

        return (
            <Tooltip text={label}>
                <button {...restProps} className={className}>
                    <span>{label}</span>
                </button>
            </Tooltip>
        );
    }
}

export default classify(defaultClasses)(Tile);
