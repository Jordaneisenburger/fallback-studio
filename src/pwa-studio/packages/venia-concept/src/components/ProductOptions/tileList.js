import React, { Component } from 'react';
import { arrayOf, object, shape, string } from 'prop-types';
import { List } from '@magento/peregrine';

import classify from '../../classify';
import Tile from './tile';
import defaultClasses from './tileList.css';

class TileList extends Component {
    static propTypes = {
        classes: shape({
            root: string
        }),
        items: arrayOf(object)
    };

    render() {
        return <List renderItem={Tile} {...this.props} />;
    }
}

export default classify(defaultClasses)(TileList);
