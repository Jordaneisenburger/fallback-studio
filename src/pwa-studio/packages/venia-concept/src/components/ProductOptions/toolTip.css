.root {
    position: relative;
}

.tooltip {
    /* Appearance. */
    --tooltip-height: 2.5rem;
    background-color: black;
    border: 1px solid #d1d1d1;
    border-radius: 2px;
    box-shadow: 0 -1px 1px 0 rgba(0, 0, 0, 0.15);
    color: white;
    height: var(--tooltip-height);
    width: max-content;

    /* Positioning. */
    position: absolute;
    top: calc((var(--tooltip-height) + 0.5rem) * -1);
    left: 0.75rem; /* Match left padding to align left. */

    /* Children positioning. */
    display: flex;
    align-items: center;
    justify-content: flex-start;
    padding: 0 0.75rem;
}

/* The tooltip triangle that points down. */
.tooltip:before {
    /* Actually make the triangle out of borders. */
    border-top: 0.5rem solid black;
    border-left: 0.5rem solid transparent;
    border-right: 0.5rem solid transparent;

    /* Triangle appearance and positioning. */
    content: '';
    display: block;
    height: 0;
    position: absolute;
    top: 100%;
    width: 0;
}

@media (max-width: 1024px) {
    .tooltip:before {
        left: calc(50% - 0.5rem);
    }
}
