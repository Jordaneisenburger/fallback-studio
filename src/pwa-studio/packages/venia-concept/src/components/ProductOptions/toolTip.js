import React, { Component } from 'react';
import { shape, string } from 'prop-types';
import over from 'lodash.over';
import uuid from 'uuid/v4';

import classify from '../../classify';
import defaultClasses from './toolTip.css';

/**
 * NOTE: The Tooltip component, as created, needs to be a parent of the
 * component it is a tooltip for. This is necessary for several reasons:
 *  - the Tooltip generates a uuid which is used for the aria-describedby
 *  - the Tooltip currently acts like a controller for event handlers.
 *
 * Once we move this component to a more general "Tooltip" component within
 * Peregrine we may change the parent-child relationship pattern. Please be
 * aware of this if you chose to reuse this component. -srugh
 */
class Tooltip extends Component {
    static propTypes = {
        classes: shape({
            root: string,
            tooltip: string
        }),
        text: string
    };

    constructor(...args) {
        super(...args);

        this.timeoutId = null;
        this.uniqueId = uuid();
    }

    state = {
        isShowing: false
    };

    componentWillUnmount() {
        if (this.timeoutId) {
            window.clearTimeout(this.timeoutId);
        }
    }

    /*
     *  Close the tooltip on the next tick.
     *  Unfortunately this is necessary because another child of ours may have received focus
     *  and this blur event is fired before that new focus event.
     *  @see https://reactjs.org/docs/accessibility.html#mouse-and-pointer-events.
     */
    onBlur = () => {
        this.timeoutId = setTimeout(() => {
            this.setState({ isShowing: false });
        }, 0);
    };

    /*
     *  The tooltip target has received focus, show the tooltip.
     */
    onFocus = () => {
        // If a child receives focus, do not close the tooltip yet.
        clearTimeout(this.timeoutId);

        this.setState({ isShowing: true });
    };

    onKeyDown = event => {
        if (event.key === 'Escape') {
            this.setState({ isShowing: false });
        }
    };
    onMouseOver = () => {
        this.setState({ isShowing: true });
    };
    onMouseLeave = () => {
        this.setState({ isShowing: false });
    };

    render() {
        const { text, children, classes } = this.props;
        const { isShowing } = this.state;

        const ariaEnhancedChildren = React.Children.map(children, child => {
            if (child.type !== 'button') {
                return child;
            }

            // The button triggers the tooltip.
            // Update its accessibility accordingly.
            const ariaEnhancedButton = React.cloneElement(child, {
                'aria-describedby': this.uniqueId,
                // We want to tack our event functionality onto whatever already exists.
                // lodash.over returns a function that calls each function
                // provided to it with the args it receives -
                // in this case, a React SyntheticEvent.
                onBlur: over([child.props.onBlur, this.onBlur]),
                onFocus: over([child.props.onFocus, this.onFocus]),
                onKeyDown: over([child.props.onKeyDown, this.onKeyDown]),
                onMouseOver: over([child.props.onMouseOver, this.onMouseOver]),
                onMouseLeave: over([
                    child.props.onMouseLeave,
                    this.onMouseLeave
                ])
            });

            return ariaEnhancedButton;
        });

        return (
            <div className={classes.root}>
                {isShowing && (
                    <div
                        className={classes.tooltip}
                        id={this.uniqueId}
                        role="tooltip"
                    >
                        {text}
                    </div>
                )}
                {ariaEnhancedChildren}
            </div>
        );
    }
}

export default classify(defaultClasses)(Tooltip);
