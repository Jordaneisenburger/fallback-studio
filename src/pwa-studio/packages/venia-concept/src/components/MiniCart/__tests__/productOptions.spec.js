import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import ProductOptions from '../productOptions';

const renderer = new ShallowRenderer();

test('it renders null when options are missing', () => {
    const tree = renderer.render(<ProductOptions />);

    expect(tree).toMatchSnapshot();
});

test('it renders null when options is empty', () => {
    const props = {
        options: []
    };
    const tree = renderer.render(<ProductOptions {...props} />);

    expect(tree).toMatchSnapshot();
});

test('it renders the correct tree', () => {
    const props = {
        options: [
            {
                label: 'Label 1',
                value: 'Value 1'
            },
            {
                label: 'Label 2',
                value: 'Value 2'
            }
        ]
    };
    const tree = renderer.render(<ProductOptions {...props} />);

    expect(tree).toMatchSnapshot();
});
