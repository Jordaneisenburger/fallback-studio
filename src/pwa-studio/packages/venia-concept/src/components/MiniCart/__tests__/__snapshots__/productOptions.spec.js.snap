// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders null when options are missing 1`] = `null`;

exports[`it renders null when options is empty 1`] = `null`;

exports[`it renders the correct tree 1`] = `
<dl>
  <React.Fragment>
    <dt>
      Label 1
       : 
      Value 1
    </dt>
  </React.Fragment>
  <React.Fragment>
    <dt>
      Label 2
       : 
      Value 2
    </dt>
  </React.Fragment>
</dl>
`;
